rasterio/_vsiopener.pyx

```diff
@@ -0,0 +1,160 @@
+cdef str PREFIX = "/vsipyopener/"
+cdef bytes PREFIX_BYTES = PREFIX.encode("utf-8")
+
+# This is global state for the Python filesystem plugin. It currently only
+# contains path -> PyOpenerBase (or subclass) instances. This is used by
+# the plugin to determine what "files" exist on "disk".
+# Currently the only way to "create" a file in the filesystem is to add
+# an entry to this dictionary. GDAL will then Open the path later.
+cdef _OPENER_REGISTRY = {}
+cdef _OPEN_FILE_OBJS = set()
+
+
+cdef int install_pyopener_plugin(VSIFilesystemPluginCallbacksStruct *callbacks_struct):
+    """Install handlers for python file openers if it isn't already installed."""
+    cdef char **registered_prefixes = VSIGetFileSystemsPrefixes()
+    cdef int prefix_index = CSLFindString(registered_prefixes, PREFIX_BYTES)
+    CSLDestroy(registered_prefixes)
+
+    if prefix_index < 0:
+        log.debug("Installing Python opener handler plugin...")
+        callbacks_struct = VSIAllocFilesystemPluginCallbacksStruct()
+        callbacks_struct.open = <VSIFilesystemPluginOpenCallback>pyopener_open
+        callbacks_struct.tell = <VSIFilesystemPluginTellCallback>pyopener_tell
+        callbacks_struct.seek = <VSIFilesystemPluginSeekCallback>pyopener_seek
+        callbacks_struct.read = <VSIFilesystemPluginReadCallback>pyopener_read
+        callbacks_struct.close = <VSIFilesystemPluginCloseCallback>pyopener_close
+        callbacks_struct.pUserData = <void*>_OPENER_REGISTRY
+        retval = VSIInstallPluginHandler(PREFIX_BYTES, callbacks_struct)
+        VSIFreeFilesystemPluginCallbacksStruct(callbacks_struct)
+        return retval
+    else:
+        return 0
+
+
+cdef void uninstall_pyopener_plugin(VSIFilesystemPluginCallbacksStruct *callbacks_struct):
+    if callbacks_struct is not NULL:
+        callbacks_struct.pUserData = NULL
+        VSIFreeFilesystemPluginCallbacksStruct(callbacks_struct)
+    callbacks_struct = NULL
+
+
+cdef void* pyopener_open(void *pUserData, const char *pszFilename, const char *pszAccess) with gil:
+    """Access files in the virtual filesystem.
+
+    This function is mandatory in the GDAL Filesystem Plugin API.
+    GDAL may call this function multiple times per filename and each
+    result must be seperately seekable.
+    """
+    if pszAccess != b"r" and pszAccess != b"rb":
+        log.error("Python opener is currently a read-only interface.")
+        return NULL
+
+    if pUserData is NULL:
+        log.error("Python opener registry is not initialized.")
+        return NULL
+
+    cdef dict registry = <object>pUserData
+    filename = pszFilename.decode("utf-8")
+
+    log.debug("Looking up opener: registry=%r, filename=%r", registry, filename)
+    try:
+        file_opener = registry[filename]
+    except KeyError:
+        log.info("Object not found: registry=%r, filename=%r", registry, filename)
+        return NULL
+
+    cdef object file_obj
+
+    try:
+        file_obj = file_opener(filename, mode="rb")
```

Comment:
The original post in this issue states that file_opener takes one positional string argument, yet we are calling it here with a keyword argument.

Reply:
Yes...

Reply:
@groutr has a good catch here. It made me wonder, what if the user needs to control the mode in their opener?
```python
>>> def open_(file, **kwargs):
...     io.open(file, mode="rb", **kwargs)
... 
>>> open_("file.tif", mode="rb")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 2, in open_
TypeError: io.open() got multiple values for keyword argument 'mode'
```

Reply:
`rasterio.open` also a mode keyword argument. How would a user disambiguate between mode for rasterio.open and mode for the opener function?

Reply:
@groutr @snowman2 I think I'm doing a poor job of explaining what's going on. `opener` is effectively called by format driver code via `VSIFOpenExL`, using a mode determined by that format (often "rb", but possibly "r" in the case of legact text-based raster formats). The user can't control the mode or the filename that is passed to `opener`. I'll try to make this more clear in the next commit.

Reply:
If VSIOpenExL is providing the mode should this be:
```
file_obj = file_opener(filename, mode=pszAccess)
```